### PR TITLE
Pojo construction from response reader

### LINKED ISSUE #82: [Code Generation] Update pojo class code generator to be constructed from ResponseReader

### PR DESCRIPTION
**Pull request by sav007**
I know this PR is quite big but I couldn't break it to smaller ones as all changes are required.

Here are couple hints:
- move guarantee unique interface name logic out from `Util` to `SchemaTypeSpecBuilder` (required for POJO constructor generation as we need to know what override name of class will be, see `HeroDetailsExpected.json` line 192 https://github.com/sav007/apollo-android/blob/0b81defb2377c2010842dbf3e672036414b878b7/apollo-compiler/src/test/graphql/com/example/pojo_fragment_with_inline_fragment/HeroDetailsExpected.java#L192)
- refactor a little bit `ResponseReader.java` to support reading list of scalar types (see `TestQueryExpected.json` line 101 https://github.com/sav007/apollo-android/blob/0b81defb2377c2010842dbf3e672036414b878b7/apollo-compiler/src/test/graphql/com/example/pojo_fragment_with_inline_fragment/TestQueryExpected.java#L101)
- change signature of `CodeGenerator` required to add `reservedTypeNames`(for class unique name logic), add `typeDeclarations` (for `SchemaTypeConstructorBuilder` to get access to declared enums as `com.squareup.javapoet.TypeName` can't tell if this is enum or not), add `abstract` (for pojo generation inside `SchemaTypeSpecBuilder`)
- move logic of pojo generation from `Util` to `SchemaTypeSpecBuilder`
- change fragment generation logic to add `TYPE_CONDITION` required by `SchemaTypeConstructorBuilder`
- add `SchemaTypeConstructorBuilder` and `SchemaFragmentsConstructorBuilder` to generate schema type construction from `ResponseReader`

Example of generated classes:
- https://github.com/sav007/apollo-android/blob/0b81defb2377c2010842dbf3e672036414b878b7/apollo-compiler/src/test/graphql/com/example/pojo_fragment_with_inline_fragment/HeroDetailsExpected.java
- https://github.com/sav007/apollo-android/blob/0b81defb2377c2010842dbf3e672036414b878b7/apollo-compiler/src/test/graphql/com/example/pojo_fragment_with_inline_fragment/TestQueryExpected.java

Closes #82

### CODE REVIEW

**Comment by felipecsl** [comment]
Wow this is a long read :) I'll take a look when I have a chance!

**Comment by digitalbuddha** [comment]
Dates are hard...how would they work?  Would passing in some sort of `DateDeserializer` be possible?

**Comment by martijnwalraven** [comment]
@digitalbuddha: I think we'd want the ability to associate your own serialization/deserialization logic with a [custom scalar type](https://facebook.github.io/graphql/#sec-Scalars).

File: apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Fragment.kt [hunk 1]
```diff
@@ -17,24 +16,34 @@ data class Fragment(
     val fragmentsReferenced: List<String>
 ) : CodeGenerator {
   /** Returns the Java interface that represents this Fragment object. */
-  override fun toTypeSpec(): TypeSpec =
-      SchemaTypeSpecBuilder().build(interfaceTypeName(), fields, fragmentSpreads, inlineFragments)
+  override fun toTypeSpec(abstract: Boolean, reservedTypeNames: List<String>,
```
Comment (felipecsl) [inline apollo-compiler/src/main/kotlin/com/apollostack/compiler/ir/Fragment.kt]: nit: `abstract` is a reserved Java keyword, maybe rename it for clarity (it also messes up the syntax highlighting)

**Comment by sav007** [comment]
@digitalbuddha @martijnwalraven #88 has been created to address custom scalar types

**Pull request merged**